rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+	Shutdown() error
+	// Module returns the module information and factory function
+	Module() core.ModuleInfo
+}
+
+// SubgraphHooks are called when a subgraph request or response is made.
+// The order is not guaranteed, so the hooks should be idempotent and side-effect free.
+// if state needs to be shared between hooks, it should be stored in the context.
+// We will provide an easy way to share state between hooks.
+type SubgraphHooks interface {
+	// OnRequest is called when a subgraph request is made.
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphRequest(req *core.SubgraphRequest, err error) error
+	// OnResponse is called when a subgraph response is received
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphResponse(res *core.SubgraphResponse, err error) error
+	// OnError is called when an error occurs during the subgraph lifecycle request/response or any other error
+	OnError(err error)
+}
+
+type ApplicationHooks interface {
+	// OnStart is called when the application starts
+	OnAppStart() error
+	// OnStop is called when the application stops
+	OnAppStop() error
+	// OnError is called when an error occurs during the application lifecycle start/stop or any other error
+	OnAppError(err error)
+}
+
+type TelemetryHooks interface {
+	// OnSpanStart is called when a span is created
+	// Returning a function to be called when the span ends.
+	// This can be used to add custom attributes or events to the span.
+	OnSpan(span *trace.Span) func() // Return a function to be called when the span ends
+}
+
+type AuthenticationHooks interface {
+	// OnAuthenticate is called when a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnAuthenticate(req *core.GatewayRequest, err error) error
+	// OnPostAuthenticate is called after a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostAuthenticate(req *core.GatewayRequest, err error) error
+}
+
+type OperationHooks interface {
+	// OnNormalize is called when an operation is normalized
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnOperationNormalize(op *core.Operation, err error) error
+	// OnPostNormalize is called after an operation is normalized
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostOperationNormalize(op *core.Operation, err error) error
+	// OnPlan is called when an operation is planned
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnOperationPlan(op *core.Operation, err error) error
+	// OnPostOperationPlan is called after an operation is planned
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostOperationPlan(op *core.Operation, err error) error
```

Comment:
These are missing:
- PreValidate (modify request before validation)
- PostValidate (add additional validation logic)
- PostParseRequestOptions (e.g. to modify trace options or execution options)
- PreReadMultipart (e.g. to handle multipart options or intercept, prevent certain uploads)
- PostReadMultipart (e.g. to modify files or do something with files, like validation)
- PreFetchPersistedOperation (e.g. to do additional validation before fetching a PO)
- PostFetchPersistedOperation (e.g. to do post processing of the PO)

Question regarding the above. Do we allow mutating state? (I think we should)
Or should we distinguish read only hooks an mutating hooks (I don't think so)

In addition, I'd also suggest to add the following hooks to "completely" override the behaviour of certain parts of the request chain.

- OverrideParse (replace existing logic with custom parse logic, implementer gets input and must return parsed operation or error)
- OverrideValidate (replace existing validation logic)
- OverrideNormalize (normalize on your own)
- OverridePlan (plan on your own)
- OverrideFetchPersistedOperation (load PO on your own)

Reply:
Most of them have been addressed. The RFC should already give a very good picture of the extensibility of the system. The current hooks are oriented on customer feedback. We can add more hooks any time.